adapters/visiblemeasures/visiblemeasures.go

```diff
@@ -0,0 +1,158 @@
+package visiblemeasures
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+type reqBodyExt struct {
+	VisibleMeasuresBidderExt reqBodyExtBidder `json:"bidder"`
+}
+
+type reqBodyExtBidder struct {
+	Type        string `json:"type"`
+	PlacementID string `json:"placementId,omitempty"`
+	EndpointID  string `json:"endpointId,omitempty"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var err error
+	var adapterRequests []*adapters.RequestData
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb2.Imp{imp}
+
+		var bidderExt adapters.ExtImpBidder
+		var visiblemeasuresExt openrtb_ext.ImpExtVisibleMeasures
+
+		if err = json.Unmarshal(reqCopy.Imp[0].Ext, &bidderExt); err != nil {
+			return nil, []error{err}
+		}
+		if err = json.Unmarshal(bidderExt.Bidder, &visiblemeasuresExt); err != nil {
+			return nil, []error{err}
+		}
+
+		temp := reqBodyExt{VisibleMeasuresBidderExt: reqBodyExtBidder{}}
+
+		if visiblemeasuresExt.PlacementID != "" {
+			temp.VisibleMeasuresBidderExt.PlacementID = visiblemeasuresExt.PlacementID
+			temp.VisibleMeasuresBidderExt.Type = "publisher"
+		} else if visiblemeasuresExt.EndpointID != "" {
+			temp.VisibleMeasuresBidderExt.EndpointID = visiblemeasuresExt.EndpointID
+			temp.VisibleMeasuresBidderExt.Type = "network"
+		}
```

Comment:
should return error if both `PlacementID` and `EndpointID` are blank

Reply:
This should not happen, because in schema `static/bidder-params/visiblemeasures.json` described that one of two is required:
`"oneOf": [
      { "required": ["placementId"] },
      { "required": ["endpointId"] }
    ]`